### PR TITLE
Added Exception for F8 Test case

### PR DESCRIPTION
**Pull request by MeritedHobbit**
Changes: 

1. Changed FW update test cases & standardized the return data. 
2. Altered F8 check to align with OCP Spec. 
3. All redfish response check messages moved to a separate input json file. 
4. Select input files, if not provided in workspace, will be picked from the json_spec/input folder. 
5. Added additional regression suite in the default test_runner.json file.

image

### CODE REVIEW

File: ctam/interfaces/fw_update_ifc.py [hunk 1]
```diff
@@ -268,6 +247,8 @@ def ctam_stage_fw(
                 StageFWOOB_Status = False
         if return_task_id:
             return StageFWOOB_Status, FwUpdTaskID
+        elif return_msg:
+            return StageFWOOB_Status, stage_msg, FwUpdTaskID
```
Comment (afsanac-nv) [inline ctam/interfaces/fw_update_ifc.py:251]: What if both return_task_id and return_msg are true? We should handle all possible scenarios.

File: ctam/interfaces/fw_update_ifc.py [hunk 2]
```diff
@@ -253,6 +224,14 @@ def ctam_stage_fw(
             else:
                 StageFWOOB_Status = True
         else:
+            error_check = JSONData.get("error", None)
+            if error_check and image_type != "large":
+                message = JSONData.get("error", {}).get("message", "")
+                if message.lower() != "An update is in progress.".lower():
```
Comment (afsanac-nv) [inline ctam/interfaces/fw_update_ifc.py:230]: Just an observation - OCP spec has this message as "Another firmware update already in progress" (Table 11), but we see "An update is in progress." in Nvidia systems as well.

File: ctam/interfaces/fw_update_ifc.py (hunk 1 above)
Comment (v-ajitpanda) [reply ctam/interfaces/fw_update_ifc.py]: This issue has been resolved.

File: ctam/interfaces/fw_update_ifc.py (hunk 2 above)
Comment (v-ajitpanda) [reply ctam/interfaces/fw_update_ifc.py]: Added one config file for redfish messages. We can add custom messages and then we can read and compare accordingly.

File: ctam/interfaces/fw_update_ifc.py [hunk 3]
```diff
@@ -650,13 +631,51 @@ def ctam_get_version_from_bundle(self, image_type):
...
+            redfish_str="{BaseURI}{TaskServiceURI}", component_type="GPU"
+        )
+        response = self.dut().run_redfish_command(uri=task_service_uri)
+        json_data = response.dict
+        task_list = [i["@odata.id"] for i in json_data["Members"]]
+        for task in task_list:
+            self.wait_for_staging(TaskID=task)
         
-    
+    def wait_for_staging(self, TaskID=None, check_time=False):
```
[the remark this answers is not in the document]
Comment (v-ajitpanda) [reply ctam/interfaces/fw_update_ifc.py]: This issue has been resolved.

**Comment by v-ajitpanda** [comment]
image

File: ctam/test_runner.py [hunk 4]
```diff
@@ -279,6 +285,7 @@ def _start(self, testrun_name="initialization"):
             logger=dut_logger,
             test_info_logger=test_info_logger,
             test_uri_response_check=self.test_uri_response_check,
```
Comment (jitendra-nvi) [inline ctam/test_runner.py:287]: @v-sourabhsingh I see that CompToolDut requires a mandatory argument `workspace_dir`.  How were the changes tested? I don't see that it's been passed here and I am facing issue while running it. Thanks.

Comment (v-sourabhsingh) [reply ctam/test_runner.py:287]: It's been fixed. Please pull latest changes.

File: ctam/interfaces/functional_ifc.py [hunk 5]
```diff
@@ -631,16 +630,23 @@ def RedfishDownloadDump(self, DumpURI):
...
+        JSONData = {}
+        if not TaskID:
+            # need to implement the flow for no task uri
+            self.test_run().add_log(LogSeverity.DEBUG,
+                    "No Task ID provided."
+                )
+            return Task_Completed, JSONData
         TaskURI = self.dut().uri_builder.format_uri(
-            redfish_str="{BaseURI}/TaskService/Tasks/" + "{}".format(TaskID), component_type="GPU"
+            redfish_str="{BaseURI}{TaskServiceURI}" + "{}".format(TaskID), component_type="GPU"
```
Comment (jitendra-nvi) [inline ctam/interfaces/functional_ifc.py:649]: @v-ajitpanda Can you please add TaskServiceURI in json spec and test your changes thoroughly? Thanks.

Comment (jitendra-nvi) [reply ctam/interfaces/functional_ifc.py:649]: Okay! Can you test the binary once for all of your changes.?

File: ctam/ctam.py [hunk 6]
```diff
@@ -117,148 +119,179 @@ def get_exception_details(exec: Exception = ""):
...
+    try:
+        # builds hierarchy of test groups and associated test cases
+        #ms_internal_tests
+
+        if args.version:
+            print(f"CTAM - version {__version__}")
+            exit()
+
+        default_config_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), "json_spec", "input")
+        print(os.path.exists(default_config_path), "FILE LOCATION")
```
Comment (jitendra-nvi) [inline ctam/ctam.py:131]: Remove debug statement.

File: ctam/ctam.py [hunk 7]
```diff
@@ -117,148 +119,179 @@ def get_exception_details(exec: Exception = ""):
...
-        test_hierarchy = TestHierarchy(ext_test_root_dir, ifc_dir)
+    try:
+        # builds hierarchy of test groups and associated test cases
+        #ms_internal_tests
+
+        if args.version:
+            print(f"CTAM - version {__version__}")
+            exit()
+
+        default_config_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), "json_spec", "input")
```
Comment (jitendra-nvi) [inline ctam/ctam.py:130]: @v-ajitpanda please add this line after 
`default_config_path = default_config_path.replace('/tmp/', '') if default_config_path.startswith('/tmp/') else default_config_path
     ` It causes error in our binary

File: json_spec/input/redfish_response_messages.json [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{
+    "UpdateProgress_Message":"UpdateInProgress",
+    "LargeFWImageUpdate": "PayloadTooLarge"
```
Comment (jitendra-nvi) [inline json_spec/input/redfish_response_messages.json:3]: @v-ajitpanda Details are missing in README.md for all new config/files. Please update README.md for all new changes.Thanks.

**Review by jitendra-nvi — CHANGES_REQUESTED** [review]

File: ctam/ctam.py (hunk 7 above)
Comment (v-ajitpanda) [reply ctam/ctam.py:130]: Added the above code

File: json_spec/input/redfish_response_messages.json (hunk 8 above)
Comment (v-ajitpanda) [reply json_spec/input/redfish_response_messages.json:3]: Removed debug statements

Comment (jitendra-nvi) [reply json_spec/input/redfish_response_messages.json:3]: @v-ajitpanda Please update the README as well. Thanks.

Comment (jitendra-nvi) [reply json_spec/input/redfish_response_messages.json:3]: Looks good, test cases passing at our end.

File: ctam/interfaces/functional_ifc.py (hunk 5 above)
Comment (MeritedHobbit) [reply ctam/interfaces/functional_ifc.py:649]: We are still having trouble with binary creation in windows. We can resolve this for the next commit.

**Pull request merged**